circuitbreaker/circuit_breaker.go

```diff
@@ -54,82 +56,81 @@ type CircuitBreaker struct {
-	resultChan := make(chan CommandResult, 1)
+// Executes the command in its circuit if set.
+// If the command's circuit is not configured, the circuit of the CircuitBreaker is used.
+// This is a blocking function.
+func (cb *CircuitBreaker) Execute(cmd Command) CommandResult {
 	var result CommandResult
+	ctx := cmd.ctx
+	if ctx == nil {
+		ctx = context.Background()
+	}
 
-	for i := 0; i < len(cmd.functors); i += 2 {
-		f1 := cmd.functors[i]
-		var f2 *Functor
-		if i+1 < len(cmd.functors) {
-			f2 = cmd.functors[i+1]
-		}
+	for i := 0; i < len(cmd.functors); i += 1 {
+		f := cmd.functors[i]
 
-		circuitName := fmt.Sprintf("%s_%d", eh.config.CommandName, i)
+		circuitName := f.circuitName
+		if circuitName == "" {
+			circuitName = cb.config.CircuitName
+		}
 		if hystrix.GetCircuitSettings()[circuitName] == nil {
 			hystrix.ConfigureCommand(circuitName, hystrix.CommandConfig{
-				Timeout:                eh.config.Timeout,
-				MaxConcurrentRequests:  eh.config.MaxConcurrentRequests,
-				RequestVolumeThreshold: eh.config.RequestVolumeThreshold,
-				SleepWindow:            eh.config.SleepWindow,
-				ErrorPercentThreshold:  eh.config.ErrorPercentThreshold,
+				Timeout:                cb.config.Timeout,
+				MaxConcurrentRequests:  cb.config.MaxConcurrentRequests,
+				RequestVolumeThreshold: cb.config.RequestVolumeThreshold,
+				SleepWindow:            cb.config.SleepWindow,
+				ErrorPercentThreshold:  cb.config.ErrorPercentThreshold,
 			})
 		}
 
-		// If circuit is the same for all functions, in case of len(cmd.functors) > 2,
-		// main and fallback providers are different next run if first two fail,
-		// which causes health issues for both main and fallback and ErrorPercentThreshold
-		// is reached faster than it should be.
-		errChan := hystrix.Go(circuitName, func() error {
-			res, err := f1.Exec()
-			// Write to resultChan only if success
-			if err == nil {
-				resultChan <- CommandResult{res: res, err: err}
-			}
-			return err
-		}, func(err error) error {
-			// In case of concurrency, we should not execute the fallback
-			if f2 == nil || err == hystrix.ErrMaxConcurrency {
-				return err
-			}
-			res, err := f2.Exec()
+		err := hystrix.DoC(ctx, circuitName, func(ctx context.Context) error {
+			res, err := f.exec()
+			// Write to result only if success
 			if err == nil {
-				resultChan <- CommandResult{res: res, err: err}
+				result = CommandResult{res: res, err: err}
 			}
 			return err
-		})
+		}, nil)
 
-		select {
-		case result = <-resultChan:
-			if result.err == nil {
-				return result
-			}
-		case err := <-errChan:
-			result = CommandResult{err: err}
+		if err == nil {
+			break
+		}
 
-			// In case of max concurrency, we should delay the execution and stop iterating over fallbacks
-			// No error unwrapping here, so use strings.Contains
-			if strings.Contains(err.Error(), hystrix.ErrMaxConcurrency.Error()) {
-				return result
-			}
+		// Accumulate errors
+		if result.err != nil {
```

Comment:
This won't be `true` ever, look at the condition in the line `110`.

Reply:
What I want to say, you have this code:
```Go
		err := hystrix.DoC(ctx, circuitName, func(ctx context.Context) error {
			res, err := f.exec()
			// Write to result only if success
			if err == nil {
				result = CommandResult{res: res, err: err}
			}
			return err
		}, nil)
```
This means you set `err` of `CommandResult` only when the error is `nil`, which is useless, cause it's nil by default.

Then in the line `124` you have this check `if result.err != nil {` which results in `false`.

Reply:
Also in the next iteration, you're not dealing with an error from that iteration, but with an error from the previous iteration cause you don't have `result.err = nil` in line `129`, which means if you have many commands you will always have in `result.err` an error from the previous iteration.

Reply:
To me, something like this, has the most sense:
```Go
		err := hystrix.DoC(ctx, circuitName, func(ctx context.Context) error {
			res, err := f.exec()
			result = CommandResult{res: res}
			return err
		}, nil)

		if err == nil {
			break
		}
		result.err = fmt.Errorf("%s.error: %w", circuitName, err)
```

Reply:
Again, it works because of this part, but entire implementation is a bit messy
```Go
		if result.err != nil {
			result.err = fmt.Errorf("%w, %s.error: %w", result.err, circuitName, err)
		} else {
			result.err = fmt.Errorf("%s.error: %w", circuitName, err)
		}
```

Reply:
For example this is simplified:
```Go
func (cb *CircuitBreaker) Execute(cmd Command) CommandResult {
	var result CommandResult
	ctx := cmd.ctx
	if ctx == nil {
		ctx = context.Background()
	}

	for i := 0; i < len(cmd.functors); i += 1 {
		f := cmd.functors[i]

		circuitName := f.circuitName
		if circuitName == "" {
			circuitName = cb.config.CircuitName
		}
		if hystrix.GetCircuitSettings()[circuitName] == nil {
			hystrix.ConfigureCommand(circuitName, hystrix.CommandConfig{
				Timeout:                cb.config.Timeout,
				MaxConcurrentRequests:  cb.config.MaxConcurrentRequests,
				RequestVolumeThreshold: cb.config.RequestVolumeThreshold,
				SleepWindow:            cb.config.SleepWindow,
				ErrorPercentThreshold:  cb.config.ErrorPercentThreshold,
			})
		}

		err := hystrix.DoC(ctx, circuitName, func(ctx context.Context) error {
			res, err := f.exec()
			result.res = res
			return err
		}, nil)

		if err == nil {
			break
		}
		result.err = fmt.Errorf("%w, %s.error: %w", result.err, circuitName, err)
	}

	return result
}
```

Reply:
Even I am not sure if we want this at all `result.err = fmt.Errorf("%w, %s.error: %w", result.err, circuitName, err)`, it's enough to be just `result.err = err`

Reply:
That's not true.
Try to use this https://github.com/status-im/status-go/pull/5457#discussion_r1663925537 and you will see.

Reply:
The changes I posted above accumulate errors.
Which test is failing?
Also I am not sure why do we want to accumulate errors at all? What's the point in that? What's the usecase for that?